scanomatic/io/jsonizer.py

```diff
@@ -245,3 +246,133 @@ def object_hook(obj: dict) -> Union[dict, Enum, Model]:
 
 def loads(s: Union[str, bytes]) -> Any:
     return json.loads(s, object_hook=object_hook)
+
+
+T = TypeVar('T')
+
+
+def copy(o: T) -> T:
+    return loads(dumps(o))
+
+
+def load(path: Union[str, Path]) -> Any:
+    if isinstance(path, str):
+        path = Path(path)
+    try:
+        return loads(path.read_text())
+    except IOError:
+        logging.warning(
+            f"Attempted to load model from '{path}', but failed",
+        )
+        return None
+
+
+def load_first(path: Union[str, Path]) -> Any:
+    content = load(path)
+    if isinstance(content, list):
+        if len(content) > 0:
+            return content[0]
+        return None
+    return content
+
+
+def _merge(model: Model, update: Model) -> bool:
+    for key in model.keys():
+        item = model[key]
+        if type(item) == type(update):
+            model[key] = update
+            return True
+        elif isinstance(item, Model):
+            if _merge(item, update):
+                return True
+    return False
+
+
+def merge_into(model: Optional[Model], update: Model) -> Model:
+    if model is None or type(model) == type(update):
+        return update
+    if not _merge(model, update):
+        logging.warning(
+            f"Attempted to update {model} with {update}, but found no matching part of the model",  # noqa: E501
+        )
+    return model
+
+
+def dump(
+    model: Model,
+    path: Union[str, Path],
+    overwrite: bool = False,
+) -> bool:
+    if overwrite:
+        model = merge_into(load(path), model)
+    try:
+        with open(path, 'w') as fh:
```

Comment:
I wonder what it says about the provenance of a programmer, whether they prefer `fp` or `fh`? I have some very unsubstantiated ideas. (I think both are fine, just curious.)

Reply:
no clue